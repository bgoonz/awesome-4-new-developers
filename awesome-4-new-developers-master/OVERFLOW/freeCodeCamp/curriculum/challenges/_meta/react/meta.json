{
  "name": "React",
  "isUpcomingChange": false,
  "dashedName": "react",
  "order": 5,
  "time": "5 hours",
  "template": "<body><div id='root'></div><div id='challenge-node'></div>${ source || '' }</body>",
  "required": [
    {
      "src": "https://unpkg.com/react@16.4.0/umd/react.production.min.js"
    },
    {
      "src": "https://unpkg.com/react-dom@16.4.0/umd/react-dom.production.min.js"
    }
  ],
  "superBlock": "front-end-development-libraries",
  "superOrder": 3,
  "challengeOrder": [
    [
      "587d7dbc367417b2b2512bb1",
      "Create a Simple JSX Element"
    ],
    [
      "5a24bbe0dba28a8d3cbd4c5d",
      "Create a Complex JSX Element"
    ],
    [
      "5a24bbe0dba28a8d3cbd4c5e",
      "Add Comments in JSX"
    ],
    [
      "5a24bbe0dba28a8d3cbd4c5f",
      "Render HTML Elements to the DOM"
    ],
    [
      "5a24c314108439a4d4036160",
      "Define an HTML Class in JSX"
    ],
    [
      "5a24c314108439a4d4036161",
      "Learn About Self-Closing JSX Tags"
    ],
    [
      "5a24c314108439a4d4036162",
      "Create a Stateless Functional Component"
    ],
    [
      "5a24c314108439a4d4036163",
      "Create a React Component"
    ],
    [
      "5a24c314108439a4d4036164",
      "Create a Component with Composition"
    ],
    [
      "5a24c314108439a4d4036165",
      "Use React to Render Nested Components"
    ],
    [
      "5a24c314108439a4d4036166",
      "Compose React Components"
    ],
    [
      "5a24c314108439a4d4036167",
      "Render a Class Component to the DOM"
    ],
    [
      "5a24c314108439a4d4036168",
      "Write a React Component from Scratch"
    ],
    [
      "5a24c314108439a4d4036169",
      "Pass Props to a Stateless Functional Component"
    ],
    [
      "5a24c314108439a4d403616a",
      "Pass an Array as Props"
    ],
    [
      "5a24c314108439a4d403616b",
      "Use Default Props"
    ],
    [
      "5a24c314108439a4d403616c",
      "Override Default Props"
    ],
    [
      "5a24c314108439a4d403616d",
      "Use PropTypes to Define the Props You Expect"
    ],
    [
      "5a24c314108439a4d403616e",
      "Access Props Using this.props"
    ],
    [
      "5a24c314108439a4d403616f",
      "Review Using Props with Stateless Functional Components"
    ],
    [
      "5a24c314108439a4d4036170",
      "Create a Stateful Component"
    ],
    [
      "5a24c314108439a4d4036171",
      "Render State in the User Interface"
    ],
    [
      "5a24c314108439a4d4036172",
      "Render State in the User Interface Another Way"
    ],
    [
      "5a24c314108439a4d4036173",
      "Set State with this.setState"
    ],
    [
      "5a24c314108439a4d4036174",
      "Bind 'this' to a Class Method"
    ],
    [
      "5a24c314108439a4d4036176",
      "Use State to Toggle an Element"
    ],
    [
      "5a24c314108439a4d4036177",
      "Write a Simple Counter"
    ],
    [
      "5a24c314108439a4d4036178",
      "Create a Controlled Input"
    ],
    [
      "5a24c314108439a4d4036179",
      "Create a Controlled Form"
    ],
    [
      "5a24c314108439a4d403617a",
      "Pass State as Props to Child Components"
    ],
    [
      "5a24c314108439a4d403617b",
      "Pass a Callback as Props"
    ],
    [
      "5a24c314108439a4d403617c",
      "Use the Lifecycle Method componentWillMount"
    ],
    [
      "5a24c314108439a4d403617d",
      "Use the Lifecycle Method componentDidMount"
    ],
    [
      "5a24c314108439a4d403617e",
      "Add Event Listeners"
    ],
    [
      "5a24c314108439a4d4036180",
      "Optimize Re-Renders with shouldComponentUpdate"
    ],
    [
      "5a24c314108439a4d4036181",
      "Introducing Inline Styles"
    ],
    [
      "5a24c314108439a4d4036182",
      "Add Inline Styles in React"
    ],
    [
      "5a24c314108439a4d4036183",
      "Use Advanced JavaScript in React Render Method"
    ],
    [
      "5a24c314108439a4d4036184",
      "Render with an If-Else Condition"
    ],
    [
      "5a24c314108439a4d4036185",
      "Use && for a More Concise Conditional"
    ],
    [
      "5a24c314108439a4d4036187",
      "Use a Ternary Expression for Conditional Rendering"
    ],
    [
      "5a24c314108439a4d4036188",
      "Render Conditionally from Props"
    ],
    [
      "5a24c314108439a4d4036189",
      "Change Inline CSS Conditionally Based on Component State"
    ],
    [
      "5a24c314108439a4d403618a",
      "Use Array.map() to Dynamically Render Elements"
    ],
    [
      "5a24c314108439a4d403618b",
      "Give Sibling Elements a Unique Key Attribute"
    ],
    [
      "5a24c314108439a4d403618c",
      "Use Array.filter() to Dynamically Filter an Array"
    ],
    [
      "5a24c314108439a4d403618d",
      "Render React on the Server with renderToString"
    ]
  ]}
