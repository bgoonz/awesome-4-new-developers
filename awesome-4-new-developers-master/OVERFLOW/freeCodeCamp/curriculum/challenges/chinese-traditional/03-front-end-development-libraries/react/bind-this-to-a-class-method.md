---
id: 5a24c314108439a4d4036174
title: 將 this 綁定到 Class 方法上
challengeType: 6
forumTopicId: 301379
dashedName: bind-this-to-a-class-method
---

# --description--

除了設置和更新 `state` 之外，還可以爲組件類定義方法。 類方法通常需要使用 `this` 關鍵字，以便它可以訪問方法中類的屬性（例如 `state` 和 `props`）。 有幾種方法可以讓類方法訪問 `this`。

一種常見的方法是在構造函數中顯式地綁定 `this`，這樣當組件初始化時，`this` 就會綁定到類方法。 你可能已經注意到上一個挑戰在構造函數中的 `handleClick` 方法使用了 `this.handleClick = this.handleClick.bind(this)`。 然後，當在類方法中調用像 `this.setState()` 這樣的函數時，`this` 指的是這個類，而不是 `undefined`。

**注意：** `this`關鍵字是 JavaScript 中最令人困惑的方面之一，但它在 React 中扮演着重要的角色。 雖然它的行爲在這裏是完全正常的，但是這些課程並不深入研究`this`，所以如果以上內容令你感到困惑，請參考其他課程！

# --instructions--

代碼編輯器有一個帶有 `state` 的組件，用於跟蹤項目計數。 它還有一個方法，允許設置文本爲 `You clicked!`。 但是，該方法不起作用，因爲它使用了未定義的 `this` 關鍵字。 可以通過將 `this` 顯式綁定到組件構造函數中的 `handleClick()`方法來修復它。

接下來，向 render 方法中的 `button` 元素添加一個單擊處理程序。 當按鈕接收到單擊事件時，它應該觸發 `handleClick()` 方法。 記住，傳遞給 `onClick` 處理程序的方法需要使用花括號，因爲它應該直接被解釋爲 JavaScript。

完成上述步驟後，可以單擊按鈕並看到 `You clicked!`。

# --hints--

`MyComponent` 應返回 `div` 元素，該元素按順序包含兩個元素，一個按鈕和一個 `h1` 元素。

```js
assert(
  Enzyme.mount(React.createElement(MyComponent)).find('div').length === 1 &&
    Enzyme.mount(React.createElement(MyComponent))
      .find('div')
      .childAt(0)
      .type() === 'button' &&
    Enzyme.mount(React.createElement(MyComponent))
      .find('div')
      .childAt(1)
      .type() === 'h1'
);
```

`MyComponent` 的 state 應該使用鍵值對 `{ text: "Hello" }`，進行初始化。

```js
assert(
  Enzyme.mount(React.createElement(MyComponent)).state('text') === 'Hello'
);
```

單擊 `button` 元素應該運行 `handleClick` 方法，並使 state `text` 爲 `You clicked!`。

```js
async () => {
  const waitForIt = (fn) =>
    new Promise((resolve, reject) => setTimeout(() => resolve(fn()), 250));
  const mockedComponent = Enzyme.mount(React.createElement(MyComponent));
  const first = () => {
    mockedComponent.setState({ text: 'Hello' });
    return waitForIt(() => mockedComponent.state('text'));
  };
  const second = () => {
    mockedComponent.find('button').simulate('click');
    return waitForIt(() => mockedComponent.state('text'));
  };
  const firstValue = await first();
  const secondValue = await second();
  assert(firstValue === 'Hello' && secondValue === 'You clicked!');
};
```

# --seed--

## --after-user-code--

```jsx
ReactDOM.render(<MyComponent />, document.getElementById('root'))
```

## --seed-contents--

```jsx
class MyComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      text: "Hello"
    };
    // Change code below this line

    // Change code above this line
  }
  handleClick() {
    this.setState({
      text: "You clicked!"
    });
  }
  render() {
    return (
      <div>
        { /* Change code below this line */ }
        <button>Click Me</button>
        { /* Change code above this line */ }
        <h1>{this.state.text}</h1>
      </div>
    );
  }
};
```

# --solutions--

```jsx
class MyComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      text: "Hello"
    };
    this.handleClick = this.handleClick.bind(this);
  }
  handleClick() {
    this.setState({
      text: "You clicked!"
    });
  }
  render() {
    return (
      <div>
        <button onClick = {this.handleClick}>Click Me</button>
        <h1>{this.state.text}</h1>
      </div>
    );
  }
};
```
