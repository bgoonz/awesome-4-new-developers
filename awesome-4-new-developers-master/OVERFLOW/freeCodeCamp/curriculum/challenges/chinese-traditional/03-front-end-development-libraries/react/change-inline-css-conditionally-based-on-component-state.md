---
id: 5a24c314108439a4d4036189
title: 根據組件狀態有條件地更改內聯 CSS
challengeType: 6
forumTopicId: 301380
dashedName: change-inline-css-conditionally-based-on-component-state
---

# --description--

此時，已經看到了一些條件渲染的應用程序和內聯樣式的使用。 這裏還有一個將這兩個主題結合在一起的例子。 你也可以根據 React 組件的 state 有條件地渲染 CSS。 要執行此操作，請檢查條件，如果滿足該條件，則修改在 render 方法中分配給 JSX 元素的樣式對象。

理解這個模式很重要，因爲相比傳統的方式（這在 jQuery 中非常常見），直接修改 DOM 元素來應用樣式的方法是一個戲劇性的轉變。 在該方法中，必須跟蹤元素何時更改並直接處理實際操作。 跟蹤更改可能變得很困難，可能會使 UI無法預測。 當根據一個條件設置一個樣式對象時，描述了 UI 作爲應用程序的狀態函數應當如何展現。 如此便有一個清晰的單向流動的信息流。 這是使用 React 編寫應用程序時的首選方法。

# --instructions--

代碼編輯器有一個簡單的帶有邊框樣式的受控 input 組件。 如果用戶在輸入框中鍵入超過 15 個字符的文本，希望將此邊框變成紅色。 添加一個條件來檢查這一點，如果條件有效，則將 input 的邊框樣式設置爲`3px solid red`。 可以通過在 input 中輸入文本來檢測它。

# --hints--

`GateKeeper` 組件應該渲染一個 `div` 元素。

```js
assert(
  (function () {
    const mockedComponent = Enzyme.mount(React.createElement(GateKeeper));
    return mockedComponent.find('div').length === 1;
  })()
);
```

`GateKeeper` 組件應使用初始爲空字符串的 `input` 進行初始化 state。

```js
assert(
  (function () {
    const mockedComponent = Enzyme.mount(React.createElement(GateKeeper));
    return mockedComponent.state().input === '';
  })()
);
```

`GateKeeper` 組件應該渲染一個 `h3` 標籤和一個 `input` 標籤。

```js
assert(
  (function () {
    const mockedComponent = Enzyme.mount(React.createElement(GateKeeper));
    return (
      mockedComponent.find('h3').length === 1 &&
      mockedComponent.find('input').length === 1
    );
  })()
);
```

`input` 標籤 `border` 屬性的樣式應該初始化爲 `1px solid black`。

```js
assert(
  (function () {
    const mockedComponent = Enzyme.mount(React.createElement(GateKeeper));
    return (
      mockedComponent.find('input').props().style.border === '1px solid black'
    );
  })()
);
```

如果 state 中 input 的值超過 15 個字符，則 `input` 標籤的 border 樣式應爲 `3px solid red`。

```js
async () => {
  const waitForIt = (fn) =>
    new Promise((resolve, reject) => setTimeout(() => resolve(fn()), 100));
  const mockedComponent = Enzyme.mount(React.createElement(GateKeeper));
  const simulateChange = (el, value) =>
    el.simulate('change', { target: { value } });
  let initialStyle = mockedComponent.find('input').props().style.border;
  const state_1 = () => {
    mockedComponent.setState({ input: 'this is 15 char' });
    return waitForIt(() => mockedComponent.find('input').props().style.border);
  };
  const state_2 = () => {
    mockedComponent.setState({
      input: 'A very long string longer than 15 characters.'
    });
    return waitForIt(() => mockedComponent.find('input').props().style.border);
  };
  const style_1 = await state_1();
  const style_2 = await state_2();
  assert(
    initialStyle === '1px solid black' &&
      style_1 === '1px solid black' &&
      style_2 === '3px solid red'
  );
};
```

# --seed--

## --after-user-code--

```jsx
ReactDOM.render(<GateKeeper />, document.getElementById('root'))
```

## --seed-contents--

```jsx
class GateKeeper extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      input: ''
    };
    this.handleChange = this.handleChange.bind(this);
  }
  handleChange(event) {
    this.setState({ input: event.target.value })
  }
  render() {
    let inputStyle = {
      border: '1px solid black'
    };
    // Change code below this line

    // Change code above this line
    return (
      <div>
        <h3>Don't Type Too Much:</h3>
        <input
          type="text"
          style={inputStyle}
          value={this.state.input}
          onChange={this.handleChange} />
      </div>
    );
  }
};
```

# --solutions--

```jsx
class GateKeeper extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      input: ''
    };
    this.handleChange = this.handleChange.bind(this);
  }
  handleChange(event) {
    this.setState({ input: event.target.value })
  }
  render() {
    let inputStyle = {
      border: '1px solid black'
    };
    if (this.state.input.length > 15) {
      inputStyle.border = '3px solid red';
    };
    return (
      <div>
        <h3>Don't Type Too Much:</h3>
        <input
          type="text"
          style={inputStyle}
          value={this.state.input}
          onChange={this.handleChange} />
      </div>
    );
  }
};
```
