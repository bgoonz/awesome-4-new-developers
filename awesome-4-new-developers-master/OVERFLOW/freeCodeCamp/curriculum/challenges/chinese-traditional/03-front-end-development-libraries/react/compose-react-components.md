---
id: 5a24c314108439a4d4036166
title: 組合 React 組件
challengeType: 6
forumTopicId: 301381
dashedName: compose-react-components
---

# --description--

隨着挑戰繼續，將組合使用更復雜的 React 組件和 JSX，有一點需要注意。 在其它組件中渲染 ES6 風格的類組件和渲染在過去幾個挑戰中使用的簡單組件沒有什麼不同。 可以在其它組件中渲染 JSX 元素、無狀態功能組件和 ES6 類組件。

# --instructions--

在代碼編輯器中，`TypesOfFood` 組件已經渲染了一個名爲 `Vegetables` 的組件。 此外，還有上次挑戰中的 `Fruits` 組件。

在 `Fruits` 中嵌套兩個組件，首先 `NonCitrus`，然後是 `Citrus`， 這兩個組件都已經引入。 接下來，將 `Fruits` 類組件嵌到 `TypesOfFood` 組件中，位於 `h1` 標題下方和 `Vegetables` 上方。 結果應該是一系列嵌套的組件，它們使用兩種不同的組件類型。

# --hints--

`TypesOfFood` 組件應該返回單個 `div` 元素。

```js
assert(
  (function () {
    const mockedComponent = Enzyme.mount(React.createElement(TypesOfFood));
    return mockedComponent.children().type() === 'div';
  })()
);
```

`TypesOfFood` 組件應該返回 `Fruits` 組件。

```js
assert(
  (function () {
    const mockedComponent = Enzyme.mount(React.createElement(TypesOfFood));
    return mockedComponent.children().childAt(1).name() === 'Fruits';
  })()
);
```

`Fruits` 組件應該返回 `NonCitrus` 組件和 `Citrus` 組件。

```js
assert(
  (function () {
    const mockedComponent = Enzyme.mount(React.createElement(TypesOfFood));
    return (
      mockedComponent.find('Fruits').children().find('NonCitrus').length ===
        1 &&
      mockedComponent.find('Fruits').children().find('Citrus').length === 1
    );
  })()
);
```

`TypesOfFood` 組件應該返回 `Vegetables` 組件，且其位於 `Fruits` 組件之下。

```js
assert(
  (function () {
    const mockedComponent = Enzyme.mount(React.createElement(TypesOfFood));
    return mockedComponent.children().childAt(2).name() === 'Vegetables';
  })()
);
```

# --seed--

## --before-user-code--

```jsx
class NonCitrus extends React.Component {
  render() {
    return (
      <div>
        <h4>Non-Citrus:</h4>
        <ul>
          <li>Apples</li>
          <li>Blueberries</li>
          <li>Strawberries</li>
          <li>Bananas</li>
        </ul>
      </div>
    );
  }
};
class Citrus extends React.Component {
  render() {
    return (
      <div>
        <h4>Citrus:</h4>
        <ul>
          <li>Lemon</li>
          <li>Lime</li>
          <li>Orange</li>
          <li>Grapefruit</li>
        </ul>
      </div>
    );
  }
};
class Vegetables extends React.Component {
  render() {
    return (
      <div>
        <h2>Vegetables:</h2>
        <ul>
          <li>Brussel Sprouts</li>
          <li>Broccoli</li>
          <li>Squash</li>
        </ul>
      </div>
    );
     }
};
```

## --after-user-code--

```jsx
ReactDOM.render(<TypesOfFood />, document.getElementById('root'))
```

## --seed-contents--

```jsx
class Fruits extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    return (
      <div>
        <h2>Fruits:</h2>
        { /* Change code below this line */ }

        { /* Change code above this line */ }
      </div>
    );
  }
};

class TypesOfFood extends React.Component {
  constructor(props) {
     super(props);
  }
  render() {
    return (
      <div>
        <h1>Types of Food:</h1>
        { /* Change code below this line */ }

        { /* Change code above this line */ }
        <Vegetables />
      </div>
    );
  }
};
```

# --solutions--

```jsx
class Fruits extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    return (
      <div>
        <h2>Fruits:</h2>
        { /* Change code below this line */ }
        <NonCitrus />
        <Citrus />
        { /* Change code above this line */ }
      </div>
    )
  }
}

class TypesOfFood extends React.Component {
  constructor(props) {
     super(props);
  }
    render() {
      return (
        <div>
        <h1>Types of Food:</h1>
          { /* Change code below this line */ }
          <Fruits />
          { /* Change code above this line */ }
          <Vegetables />
        </div>
      );
    }
};
```
