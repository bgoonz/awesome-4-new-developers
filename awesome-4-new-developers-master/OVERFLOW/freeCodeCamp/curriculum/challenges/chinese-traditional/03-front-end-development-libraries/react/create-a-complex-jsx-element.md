---
id: 5a24bbe0dba28a8d3cbd4c5d
title: 創建一個複雜的 JSX 元素
challengeType: 6
forumTopicId: 301382
dashedName: create-a-complex-jsx-element
---

# --description--

上一個挑戰是 JSX 的一個簡單示例，但 JSX 也可以表示更復雜的 HTML。

關於嵌套的 JSX，需要知道的一件重要的事情，那就是它必須返回單個元素。

這個父元素將包裹所有其他級別的嵌套元素。

例如，幾個作爲兄弟元素編寫的 JSX 元素而沒有父元素包裹將不會被轉換。

這裏是一個示例：

**有效的 JSX：**

```jsx
<div>
  <p>Paragraph One</p>
  <p>Paragraph Two</p>
  <p>Paragraph Three</p>
</div>
```

**無效的 JSX：**

```jsx
<p>Paragraph One</p>
<p>Paragraph Two</p>
<p>Paragraph Three</p>
```

# --instructions--

定義一個新的常量 `JSX`，渲染一個 `div`，其中依次包含以下元素：

一個 `h1`，一個 `p`，一個包含三個 `li` 項的無序列表。 可以在每個元素中包含任意文本。

**注意：** 當像這樣渲染多個元素時，可以把它們都用圓括號括起來，但是這並不是必須的。 另外，此挑戰使用 `div` 標籤把所有子元素包裹在裏面。 如果刪除 `div`，JSX 將不會編譯這些元素。 請記住這一點，因爲在 React 組件中返回 JSX 元素時也適用。

# --hints--

常量 `JSX` 應該返回一個 `div` 元素。

```js
assert(JSX.type === 'div');
```

`div` 應該包含一個 `h1` 標籤作爲第一個元素。

```js
assert(JSX.props.children[0].type === 'h1');
```

`div`應該包含一個`p`標籤作爲第二個元素。

```js
assert(JSX.props.children[1].type === 'p');
```

`div` 應該包含一個 `ul` 標籤作爲第三個元素。

```js
assert(JSX.props.children[2].type === 'ul');
```

`ul` 應該包含三個 `li` 元素。

```js
assert(
  JSX.props.children
    .filter((ele) => ele.type === 'ul')[0]
    .props.children.filter((ele) => ele.type === 'li').length === 3
);
```

# --seed--

## --after-user-code--

```jsx
ReactDOM.render(JSX, document.getElementById('root'))
```

## --seed-contents--

```jsx

```

# --solutions--

```jsx
const JSX = (
<div>
  <h1>Hello JSX!</h1>
  <p>Some info</p>
  <ul>
    <li>An item</li>
    <li>Another item</li>
    <li>A third item</li>
  </ul>
</div>);
```
