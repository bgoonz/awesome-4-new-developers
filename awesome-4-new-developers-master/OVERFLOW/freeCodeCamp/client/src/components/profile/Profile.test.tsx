import { render, screen } from '@testing-library/react';
import React from 'react';

import Profile from './Profile';

jest.mock('../../analytics');

const userProps = {
  user: {
    profileUI: {
      isLocked: false,
      showAbout: false,
      showCerts: false,
      showHeatMap: false,
      showLocation: false,
      showName: false,
      showPoints: false,
      showPortfolio: false,
      showTimeLine: false,
      showDonation: false
    },
    calendar: {},
    streak: {
      current: 1,
      longest: 1
    },
    completedChallenges: [],
    portfolio: [],
    about: 'string',
    githubProfile: 'string',
    isGithub: true,
    isLinkedIn: true,
    isTwitter: true,
    isWebsite: true,
    joinDate: 'string',
    linkedin: 'string',
    location: 'string',
    name: 'string',
    picture: 'string',
    points: 1,
    twitter: 'string',
    username: 'string',
    website: 'string',
    yearsTopContributor: [],
    isDonating: false
  },
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  navigate: () => {}
};

const notMyProfileProps = {
  isSessionUser: false,
  ...userProps
};

describe('<Profile/>', () => {
  it('renders the report button on another persons profile', () => {
    render(<Profile {...notMyProfileProps} />);

    const reportButton: HTMLElement = screen.getByText('buttons.flag-user');
    expect(reportButton).toHaveAttribute('href', '/user/string/report-user');
  });

  it('renders correctly', () => {
    const { container } = render(<Profile {...notMyProfileProps} />);

    expect(container).toMatchSnapshot();
  });
});
