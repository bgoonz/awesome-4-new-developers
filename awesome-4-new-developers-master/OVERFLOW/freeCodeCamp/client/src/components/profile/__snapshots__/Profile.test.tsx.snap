// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile/> renders correctly 1`] = `
<div>
  <div
    class="spacer"
    style="padding: 15px 0px; height: 1px;"
  />
  <div
    class="container"
  >
    <div
      class="spacer"
      style="padding: 15px 0px; height: 1px;"
    />
    <div>
      <div
        class="row"
      >
        <div
          class="avatar-camper col-xs-12"
        >
          <div
            class="avatar-container default-border"
          >
            <img
              alt="profile.avatar"
              class="avatar img-responsive"
              src="string"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="text-center social-media-icons col-sm-6 col-sm-offset-3"
        >
          <a
            aria-label="aria.linkedin"
            href="string"
            rel="noopener noreferrer"
            target="_blank"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-linkedin fa-w-14 fa-2x "
              data-icon="linkedin"
              data-prefix="fab"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"
                fill="currentColor"
              />
            </svg>
          </a>
          <a
            aria-label="aria.github"
            href="string"
            rel="noopener noreferrer"
            target="_blank"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-github fa-w-16 fa-2x "
              data-icon="github"
              data-prefix="fab"
              focusable="false"
              role="img"
              viewBox="0 0 496 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
                fill="currentColor"
              />
            </svg>
          </a>
          <a
            aria-label="aria.website"
            href="string"
            rel="noopener noreferrer"
            target="_blank"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-link fa-w-16 fa-2x "
              data-icon="link"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
                fill="currentColor"
              />
            </svg>
          </a>
          <a
            aria-label="aria.twitter"
            href="string"
            rel="noopener noreferrer"
            target="_blank"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-twitter fa-w-16 fa-2x "
              data-icon="twitter"
              data-prefix="fab"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
                fill="currentColor"
              />
            </svg>
          </a>
        </div>
      </div>
      <br />
      <h2
        class="text-center username"
      >
        @
        string
      </h2>
      
      
      
      
      <br />
    </div>
    <div
      class="spacer"
      style="padding: 15px 0px; height: 1px;"
    />
    <div
      class="text-center row"
    >
      <a
        href="/user/string/report-user"
      >
        buttons.flag-user
      </a>
    </div>
    <div
      class="spacer"
      style="padding: 15px 0px; height: 1px;"
    />
  </div>
</div>
`;
